### PR TITLE
feat: brightness prioritization

### LINKED ISSUE #588: `separate_turn_on_commands`: Adjust brightness first?
I would like to propose a user experience improvement to `separate_turn_on_commands`. Currently, when a command is split into two, the first command updates the color temperature, and the second one the brightness. I think that's not always ideal.

Here's an example (let's assume the `initial_transition` is set to `2` seconds):

1. Turn on lights during the day. AL sets brightness to 100%.
2. Turn off lights during the day.
3. Later during the night, you wake up and wander from your bedroom to your bathroom.
4. You turn on the bedroom lights. They first ramp up to 100% and burn your eyes for one second while AL adjusts the color temperature. Only afterwards, AL dims them down to 1%.
5. Relieved, you cross the hallway, where the same thing happens.
6. Then again in the bathroom. You're fully awake 😬😅 

So at night, I'd rather have a dimmed light at a wrong color temperature (that adjusts with a little delay) than a shockingly bright light. During the day, I think ramping up the brightness first might also be the smarter choice, because you probably don't want to wait 1 second until the light gets bright enough to be useful. Overall, a correct brightness settings could be more important for the utility of a light than the color temperature, which tends to be more of a cosmetic function.

There are multiple factors determining if and how bad this really is perceived, e.g., the distance of the starting brightness/temperature to the desired value, the configuration (transition length, min/max brightness, ...), used hardware, user sensitivity, and the example describes a worst-case scenario. But overall, I think this would be an optimization for the more extreme cases while not negatively affecting lighter cases.

As for the implementation, I think it's just a reorder of the two `if` blocks here: https://github.com/basnijholt/adaptive-lighting/blob/9caf64509349958a340f8cef9207fdb0ea35dbef/custom_components/adaptive_lighting/switch.py#L290-L298

Bonus improvement affecting the same code block: Skip unnecessary commands where lights are already at the desired state (e.g., to minimize Zigbee traffic).

What do you think?

### PR DESCRIPTION
**Pull request by protyposis**
Change the adaptation sequence of the `separate_turn_on_commands` mode from color/brightness to brightness/color. Fixes #588.

Includes a refactoring of the logic that executes the adaptation service calls with the following advantages:
* The same code path is executed, independent of `separate_turn_on_commands`.
* Refactors `_split_service_data` into a more generic `_prepare_service_calls` which isn't hardcoded to splitting into two service calls. Also serves as a preparation for #589.
* Improves #590 (decreases the likelihood of issue #410) by moving all adaptation commands into the cancellable task, even when `separate_turn_on_commands` is disabled.

### CODE REVIEW

**Comment by basnijholt** [comment]
Great work! 😃❤️

Do you think it makes sense to just make this the default? And can you think of any sensible use case where someone wants to adapt the color first?

**Comment by basnijholt** [comment]
I probably didn't think about it well enough when writing that code initially. Yes, I would vote for removing the configuration option and just flip the order 🙂

**Comment by protyposis** [comment]
Updated 👍 Due to the removal of the flag, the refactoring became unnecessary, but I think it's still valuable to keep for the reasons listed.

**Comment by protyposis** [comment]
Seems like this was a timing problem due to all service calls now being invoked asynchronously, and the test works again. I'm not entirely sure whether this is a proper fix, though. The remaining failing tests on with the HA dev branch also fail on the main branch.

File: tests/test_switch.py [hunk 1]
```diff
@@ -777,8 +778,8 @@ async def turn_light(state, **kwargs):
     assert (
         switch.extra_state_attributes["autoreset_time_remaining"][light.entity_id] > 0
     )
-    await asyncio.sleep(0.3)  # Should be enough time for auto reset
     await update()
+    await asyncio.sleep(0.3)  # Should be enough time for auto reset
```
Comment (basnijholt) [inline tests/test_switch.py:782]: Why did you move this statement?

File: tests/test_switch.py [hunk 2]
```diff
@@ -1385,3 +1386,105 @@ async def change_switch_settings(**kwargs):
     # testing with "configuration" should revert back to 2500
     await change_switch_settings(**{CONF_USE_DEFAULTS: "configuration"})
     assert switch._sun_light_settings.min_color_temp == 2500
+
+
+@pytest.mark.parametrize(
```
Comment (basnijholt) [inline tests/test_switch.py:1391]: Awesome, thanks for adding tests!

File: tests/test_switch.py (hunk 1 above)
Comment (protyposis) [reply tests/test_switch.py:782]: This fixes the broken test (what I tried to explain in https://github.com/basnijholt/adaptive-lighting/pull/598#issuecomment-1566170628).

Comment (basnijholt) [reply tests/test_switch.py:782]: This bit is the only reason I haven't clicked merge yet. I want to understand why this is required and I am wondering whether this indicates whether anything broke.

Comment (protyposis) [reply tests/test_switch.py:782]: I think the reason this is required is that the service call is now executed from the event loop (due to now being `async`), not directly/synchronously as it used to be. So I could have kept the `sleep(0.3)` where it was and added another `sleep` here to await processing of the event loop. Simply moving it down seemed simpler. But like I said, I'm not entirely sure because I'm far from being a Python expert :)

Comment (protyposis) [reply tests/test_switch.py:782]: So what's the status, did my answer help in any way? Whether you merge this or not is both fine with me, but it would be helpful if I knew what to expect.

Comment (basnijholt) [reply tests/test_switch.py:782]: I’ll just go on good faith and merge 😃

Thanks a lot for your contribution!
